File: scripts/doctor.py
from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path
from typing import Any

from resolve_adapter import load_adapter

COMMAND_FIELDS = (
    "iterate_command_templates",
    "held_out_command_templates",
    "comparison_command_templates",
    "full_gate_command_templates",
)


def _has_items(value: Any) -> bool:
    return isinstance(value, list) and len(value) > 0


def _check(id: str, ok: bool, detail: str) -> dict[str, Any]:
    return {"id": id, "ok": ok, "detail": detail}


def _init_command(repo_root: Path, adapter_name: str | None) -> str:
    command = f"node ./bin/cautilus adapter init --repo-root {repo_root}"
    if adapter_name:
        command += f" --adapter-name {adapter_name}"
    return command


def doctor_repo(
    repo_root: Path,
    adapter: Path | None = None,
    adapter_name: str | None = None,
) -> dict[str, Any]:
    payload = load_adapter(repo_root, adapter=adapter, adapter_name=adapter_name)
    checks: list[dict[str, Any]] = []
    suggestions: list[str] = []
    data = payload.get("data", {})

    if not payload["found"]:
        checks.append(_check("adapter_found", False, "No checked-in adapter was found."))
        status = "missing_adapter"
        summary = "Adapter missing."
        suggestions.append(_init_command(repo_root, adapter_name))
        if payload.get("warnings"):
            suggestions.extend(payload["warnings"])
        return {
            "status": status,
            "ready": False,
            "summary": summary,
            "repo_root": str(repo_root),
            "adapter_path": payload.get("path"),
            "searched_paths": payload.get("searched_paths", []),
            "checks": checks,
            "suggestions": suggestions,
            "warnings": payload.get("warnings", []),
            "errors": payload.get("errors", []),
        }

    checks.append(_check("adapter_found", True, f"Using adapter at {payload['path']}"))

    if not payload["valid"]:
        checks.append(_check("adapter_valid", False, "Adapter failed schema validation."))
        suggestions.append("Repair the adapter fields reported in errors before running evaluation.")
        suggestions.append("See docs/contracts/adapter-contract.md for the canonical adapter shape.")
        return {
            "status": "invalid_adapter",
            "ready": False,
            "summary": "Adapter is present but invalid.",
            "repo_root": str(repo_root),
            "adapter_path": payload.get("path"),
            "searched_paths": payload.get("searched_paths", []),
            "checks": checks,
            "suggestions": suggestions,
            "warnings": payload.get("warnings", []),
            "errors": payload.get("errors", []),
        }

    checks.append(_check("adapter_valid", True, "Adapter passed schema validation."))

    has_repo = isinstance(data.get("repo"), str) and bool(data["repo"].strip())
    checks.append(
        _check(
            "repo_name",
            has_repo,
            "Adapter declares repo." if has_repo else "Adapter is missing a repo name.",
        )
    )
    if not has_repo:
        suggestions.append("Set adapter.repo to the host repo name.")

    has_surfaces = _has_items(data.get("evaluation_surfaces"))
    checks.append(
        _check(
            "evaluation_surfaces",
            has_surfaces,
            "Adapter declares evaluation surfaces."
            if has_surfaces
            else "Adapter is missing evaluation_surfaces.",
        )
    )
    if not has_surfaces:
        suggestions.append("Add at least one evaluation_surfaces entry that states what the adapter judges.")

    has_baselines = _has_items(data.get("baseline_options"))
    checks.append(
        _check(
            "baseline_options",
            has_baselines,
            "Adapter declares baseline options."
            if has_baselines
            else "Adapter is missing baseline_options.",
        )
    )
    if not has_baselines:
        suggestions.append("Add at least one baseline_options entry so comparisons stay explicit.")

    automated_commands = any(_has_items(data.get(field)) for field in COMMAND_FIELDS)
    has_variants = _has_items(data.get("executor_variants"))
    has_execution_surface = automated_commands or has_variants
    checks.append(
        _check(
            "execution_surface",
            has_execution_surface,
            "Adapter declares runnable command templates or executor variants."
            if has_execution_surface
            else "Adapter has no command templates or executor variants yet.",
        )
    )
    if not has_execution_surface:
        suggestions.append(
            "Add at least one iterate/held_out/comparison/full_gate command template or executor_variants entry."
        )

    ready = all(check["ok"] for check in checks)
    summary = "Adapter is ready for standalone Cautilus use." if ready else "Adapter exists but is not ready yet."
    return {
        "status": "ready" if ready else "incomplete_adapter",
        "ready": ready,
        "summary": summary,
        "repo_root": str(repo_root),
        "adapter_path": payload.get("path"),
        "searched_paths": payload.get("searched_paths", []),
        "checks": checks,
        "suggestions": suggestions,
        "warnings": payload.get("warnings", []),
        "errors": payload.get("errors", []),
    }


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--repo-root", type=Path, required=True)
    parser.add_argument("--adapter", type=Path)
    parser.add_argument("--adapter-name")
    args = parser.parse_args()
    result = doctor_repo(args.repo_root.resolve(), adapter=args.adapter, adapter_name=args.adapter_name)
    sys.stdout.write(json.dumps(result, ensure_ascii=False, indent=2, sort_keys=True) + "\n")
    raise SystemExit(0 if result["ready"] else 1)


if __name__ == "__main__":
    main()

File: scripts/doctor.mjs
import process from "node:process";
import { pathToFileURL } from "node:url";
import { resolve } from "node:path";

import { loadAdapter } from "./resolve_adapter.mjs";

const COMMAND_FIELDS = [
	"eval_test_command_templates",
	"iterate_command_templates",
	"held_out_command_templates",
	"comparison_command_templates",
	"full_gate_command_templates",
];

function hasItems(value) {
	return Array.isArray(value) && value.length > 0;
}

function check(id, ok, detail) {
	return { id, ok, detail };
}

function initCommand(repoRoot, adapterName) {
	let command = `cautilus adapter init --repo-root ${repoRoot}`;
	if (adapterName) {
		command += ` --adapter-name ${adapterName}`;
	}
	return command;
}

function baseResult(repoRoot, payload, checks, suggestions) {
	return {
		repo_root: repoRoot,
		adapter_path: payload.path,
		searched_paths: payload.searched_paths ?? [],
		checks,
		suggestions,
		warnings: payload.warnings ?? [],
		errors: payload.errors ?? [],
	};
}

function missingAdapterResult(repoRoot, payload, adapterName) {
	const checks = [check("adapter_found", false, "No checked-in adapter was found.")];
	const suggestions = [initCommand(repoRoot, adapterName)];
	if (payload.warnings?.length) {
		suggestions.push(...payload.warnings);
	}
	return {
		status: "missing_adapter",
		ready: false,
		summary: "Adapter missing.",
		...baseResult(repoRoot, payload, checks, suggestions),
	};
}

function invalidAdapterResult(repoRoot, payload, checks) {
	return {
		status: "invalid_adapter",
		ready: false,
		summary: "Adapter is present but invalid.",
		...baseResult(repoRoot, payload, checks, [
			"Repair the adapter fields reported in errors before running evaluation.",
			"See docs/contracts/adapter-contract.md for the canonical adapter shape.",
		]),
	};
}

function pushFieldCheck(checks, suggestions, { id, ok, okDetail, missingDetail, suggestion }) {
	checks.push(check(id, ok, ok ? okDetail : missingDetail));
	if (!ok) {
		suggestions.push(suggestion);
	}
}

function collectReadinessChecks(data, checks, suggestions) {
	const hasRepo = typeof data.repo === "string" && Boolean(data.repo.trim());
	pushFieldCheck(checks, suggestions, {
		id: "repo_name",
		ok: hasRepo,
		okDetail: "Adapter declares repo.",
		missingDetail: "Adapter is missing a repo name.",
		suggestion: "Set adapter.repo to the host repo name.",
	});

	const hasSurfaces = hasItems(data.evaluation_surfaces);
	pushFieldCheck(checks, suggestions, {
		id: "evaluation_surfaces",
		ok: hasSurfaces,
		okDetail: "Adapter declares evaluation surfaces.",
		missingDetail: "Adapter is missing evaluation_surfaces.",
		suggestion: "Add at least one evaluation_surfaces entry that states what the adapter judges.",
	});

	const hasBaselines = hasItems(data.baseline_options);
	pushFieldCheck(checks, suggestions, {
		id: "baseline_options",
		ok: hasBaselines,
		okDetail: "Adapter declares baseline options.",
		missingDetail: "Adapter is missing baseline_options.",
		suggestion: "Add at least one baseline_options entry so comparisons stay explicit.",
	});

	const automatedCommands = COMMAND_FIELDS.some((field) => hasItems(data[field]));
	const hasVariants = hasItems(data.executor_variants);
	pushFieldCheck(checks, suggestions, {
		id: "execution_surface",
		ok: automatedCommands || hasVariants,
		okDetail: "Adapter declares runnable command templates or executor variants.",
		missingDetail: "Adapter has no command templates or executor variants yet.",
		suggestion:
			"Add at least one skill_test/iterate/held_out/comparison/full_gate command template or executor_variants entry.",
	});
}

export function doctorRepo(repoRoot, { adapter = null, adapterName = null } = {}) {
	const payload = loadAdapter(repoRoot, { adapter, adapterName });
	const checks = [];
	const suggestions = [];
	const data = payload.data ?? {};

	if (!payload.found) {
		return missingAdapterResult(repoRoot, payload, adapterName);
	}

	checks.push(check("adapter_found", true, `Using adapter at ${payload.path}`));

	if (!payload.valid) {
		checks.push(check("adapter_valid", false, "Adapter failed schema validation."));
		return invalidAdapterResult(repoRoot, payload, checks);
	}

	checks.push(check("adapter_valid", true, "Adapter passed schema validation."));
	collectReadinessChecks(data, checks, suggestions);

	const ready = checks.every((item) => item.ok);
	return {
		status: ready ? "ready" : "incomplete_adapter",
		ready,
		summary: ready ? "Adapter is ready for standalone Cautilus use." : "Adapter exists but is not ready yet.",
		...baseResult(repoRoot, payload, checks, suggestions),
	};
}

function fail(message) {
	process.stderr.write(`${message}\n`);
	process.exit(1);
}

function usage(exitCode = 0) {
	const text = [
		"Usage:",
		"  node ./scripts/doctor.mjs --repo-root <dir> [--adapter <path>] [--adapter-name <name>]",
	].join("\n");
	const out = exitCode === 0 ? process.stdout : process.stderr;
	out.write(`${text}\n`);
	process.exit(exitCode);
}

function parseArgs(argv) {
	const options = {
		repoRoot: null,
		adapter: null,
		adapterName: null,
	};
	for (let index = 0; index < argv.length; index += 1) {
		const arg = argv[index];
		if (arg === "-h" || arg === "--help") {
			usage(0);
		}
		const field = {
			"--repo-root": "repoRoot",
			"--adapter": "adapter",
			"--adapter-name": "adapterName",
		}[arg];
		if (!field) {
			fail(`Unknown argument: ${arg}`);
		}
		const value = argv[index + 1];
		if (!value) {
			fail(`Missing value for ${arg}`);
		}
		options[field] = value;
		index += 1;
	}
	if (!options.repoRoot) {
		fail("--repo-root is required");
	}
	if (options.adapter && options.adapterName) {
		fail("Use either --adapter or --adapter-name, not both.");
	}
	return options;
}

export function main(argv = process.argv.slice(2)) {
	const options = parseArgs(argv);
	const result = doctorRepo(resolve(options.repoRoot), {
		adapter: options.adapter,
		adapterName: options.adapterName,
	});
	process.stdout.write(`${JSON.stringify(result, null, 2)}\n`);
	process.exit(result.ready ? 0 : 1);
}

const entryHref = process.argv[1] ? pathToFileURL(resolve(process.argv[1])).href : "";
if (import.meta.url === entryHref) {
	main();
}
